### ISSUE TITLE
SpatialAttention复现出现梯度归零警报

### ISSUE THREAD

**Issue by fenghansen** [MegEngine/MegEngine#108]
##环境
1.系统环境：MegStudio
2.MegEngine版本：default
3.python版本：default

## 复现步骤
旷视视频超分比赛，将torch实现的CBAM模块迁移到megengine过程中出现了警报。
我定位了一下应该是SpatialAttention的问题.
我把torch.max(x,dim=1,keepdim=True)和torch.min(x,dim=1,keepdim=True)改成了对应天元函数的F.max(x,dim=1,keepdims=True)和F.mean(x,dim=1,keepdims=True)【考虑到一个叫keepdim一个叫keepdims，我打算直接传参数(x,1,True)】

## 目前的代码
`use_mge=True
class SpatialAttention(nn.Module):
    def __init__(self, kernel_size=3):
        super().__init__()
        self.conv = nn.Conv2d(2,1,kernel_size, padding=1, bias=False)
        self.sigmoid = nn.Sigmoid()
        self.concat = Concat()
        self.mean = F.mean if use_mge else torch.mean
        self.max = F.max if use_mge else torch.max

    def forward(self, x):
        avgout = self.mean(x, 1, keepdim=True)
        maxout, _ = self.max(x, 1, keepdim=True)
        x = self.concat([avgout, maxout], 1)
        x = self.conv(x)
        return self.sigmoid(x)
`

## 请提供完整的日志及报错信息
这是个Warning，我不太明白这个警报的具体意义，只是没生效而已还是梯度在此中断了呢？
`17 05:13:23[mgb] WRN target node ({id:15400, shape:{1}, Float32, owner:TRUE_DIV(reduce0[15382],asFloat32[15390])[15399]{Elemwise}, name:TRUE_DIV(reduce0[15382],asFloat32[15390])[15399], slot:0, gpu0:0, s, 2, 8}) does not depend on wrt node ({id:2909, shape:{1,2,3,3}, Float32, owner:shared[2908]{VolatileSharedDeviceTensor}, name:shared[2908], slot:0, gpu0:0, s, 4, 1}), return zeros as grad
17 05:13:23[mgb] WRN target node ({id:15400, shape:{1}, Float32, owner:TRUE_DIV(reduce0[15382],asFloat32[15390])[15399]{Elemwise}, name:TRUE_DIV(reduce0[15382],asFloat32[15390])[15399], slot:0, gpu0:0, s, 2, 8}) does not depend on wrt node ({id:2903, shape:{1,2,3,3}, Float32, owner:shared[2902]{VolatileSharedDeviceTensor}, name:shared[2902], slot:0, gpu0:0, s, 4, 1}), return zeros as grad`

**Comment by fenghansen** [comment]
传错代码版本了，抱歉，重新传一下代码
`class SpatialAttention(nn.Module):
    def __init__(self, kernel_size=3):
        super().__init__()
        self.conv = nn.Conv2d(2,1,kernel_size, padding=1, bias=False)
        self.sigmoid = nn.Sigmoid()
        self.concat = Concat()
        self.mean = F.mean if use_mge else torch.mean
        self.max = F.max if use_mge else torch.max

    def forward(self, x):
        avgout = self.mean(x, 1, True)
        maxout = self.max(x, 1, True)
        x = self.concat([avgout, maxout], 1)
        x = self.conv(x)
        return self.sigmoid(x)`

**Comment by fenghansen** [comment]
传错代码版本了，是下面这个版本：
`
class SpatialAttention(nn.Module):
    def __init__(self, kernel_size=3):
        super().__init__()
        self.conv = nn.Conv2d(2,1,kernel_size, padding=1, bias=False)
        self.sigmoid = nn.Sigmoid()
        self.concat = Concat()
        self.mean = F.mean if use_mge else torch.mean
        self.max = F.max if use_mge else torch.max

    def forward(self, x):
        avgout = self.mean(x, 1, True)
        maxout = self.max(x, 1, True)
        x = self.concat([avgout, maxout], 1)
        x = self.conv(x)
        return self.sigmoid(x)
‘

**Comment by qq332982511** [comment]
这个问题不大好看，如果能提供最小复现的python脚本就好了。可以对你的网络一顿删，只保留能复现问题的部分就行，数据用随机数复现更好。

**Comment by lqy-megvii** [comment]
你好，我根据你的 Module 代码，写了如下脚本，看起来可以正常训练：
```
import numpy as np

import megengine as mge
import megengine.functional as F
import megengine.module as M
import megengine.optimizer as optim


class SpatialAttention(M.Module):
    def __init__(self, kernel_size=3):
        super().__init__()
        self.conv = M.Conv2d(2, 1, kernel_size, padding=1, bias=False)
        self.sigmoid = M.Sigmoid()
        self.concat = M.Concat()
        self.mean = F.mean 
        self.max = F.max 

    def forward(self, x):
        avgout = self.mean(x, 1, True)
        maxout = self.max(x, 1, True)
        x = self.concat([avgout, maxout], 1)
        x = self.conv(x)
        return self.sigmoid(x)


if __name__ == "__main__":
    data = mge.tensor(np.random.random((1, 2, 10, 10)).astype(np.float32))
    model = SpatialAttention()
    opt = optim.SGD(model.parameters(), lr=0.1)

    for i in range(5):
        opt.zero_grad()
        loss = model(data).mean()
        opt.backward(loss)
        opt.step()
        print("loss = {:.3f}".format(loss.numpy()[0]))
```

**Comment by lqy-megvii** [comment]
如果你仍然遇到 grad = 0 的问题，请提供一个完整的复现代码

**Comment by fenghansen** [comment]
感谢你的答复让我发现我犯了一个非常低级的错误（return错变量了……）
我为我所耽误了你的时间而感到十分抱歉。
